recipe/patches/0003-Fix-fileobj-I-O-un-deterministic-behavior.patch

```diff
@@ -0,0 +1,86 @@
+From df8b1e33705e45d0e70af794d819d80e4772a2b2 Mon Sep 17 00:00:00 2001
+From: moto <855818+mthrok@users.noreply.github.com>
+Date: Tue, 23 Feb 2021 13:08:53 -0500
+Subject: [PATCH 3/3] Fix fileobj I/O un-deterministic behavior
+MIME-Version: 1.0
+Content-Type: text/plain; charset=UTF-8
+Content-Transfer-Encoding: 8bit
+
+Description taken from https://github.com/pytorch/audio/pull/1297
+
+> ## What was the cause and what is the fix?
+>
+> When opening file (for both read and write), `libsox` initializes the
+> dedicated data structure called `sox_format_t`, and set various attributes.
+> One of which is `seekable` and this value is supposed to be `false` for
+> file-like objects. In the investigation #1229, it turned out that this
+> attribute is occasionally `true` which contributed to the un-deterministic
+> behavior described above. The following code from `libsox` determines if the
+> opened `FILE*` is `seekable`, by simply checking if the associated file
+> descriptor is a regular file.
+> [[src](https://github.com/dmkrepo/libsox/blob/master/src/formats.c#L329-L338)]
+>
+> ```c
+> static sox_bool is_seekable(sox_format_t const * ft)
+> {
+>   struct stat st;
+>
+>   assert(ft);
+>   if (!ft->fp)
+>     return sox_false;
+>   fstat(fileno((FILE*)ft->fp), &st);
+>   return ((st.st_mode & S_IFMT) == S_IFREG);
+> }
+> ```
+>
+> The problem is that `fileno` returns `-1` for the case of file-like object.
+> This is because for file-like object, we use [`sox_open_mem_read`](<repo>/src/formats.c#L617)
+> or [`sox_open_mem_stream_write`](<repo>/src/formats.c#L991), which calls
+> [`fmemopen`](<repo>/src/formats.c#L496) and [`open_memstream`](<repo>/src/formats.c#L894)
+> respectively, and the resulting `*FILE` object does not have the associated
+> file descriptor. In this case, `fstat` received `-1` as the file descriptor
+> value and (I guess) the `stat` structure is untouched. Thus the attribute of
+> `st_mode` has an un-initialized value, which result in un-deterministic behavior.
+>
+> This PR adds patch to check the the value of file descriptor and mark
+> `seekable=false` if it does not have a valid file descriptor.
+
+[...]
+
+> ## Upstream situation
+>
+> ⚠️ the upstream has a fix https://sourceforge.net/p/sox/code/ci/bb38934e11035c8fab141f70dabda3afdd17da36/,
+> but this fix seems to make `is_seekable` return `true` for the in-memory file
+> object case (need to verify this), which is the opposite of the behavior we
+> want in torchaudio implementation.
```

Comment:
What's the issue for torchaudio with seekable in-memory file objects?

Reply:
Hey @mmcauliffe, thanks for the quick response!

> What's the issue for torchaudio with seekable in-memory file objects?

We'd have to ask @mthrok who did the debugging/digging/fixing in https://github.com/pytorch/audio/pull/1297 ff.

Reply:
It's described in the upstream patch, the original implementation `fstat(fileno((FILE*)ft->fp), &st)` did not check the return value of `fileno`. and the `fstat` is failing but ignored, as a result `st.st_mode` attribute is not initialized and has random value. So it caused inconsistent behavior.
 
I think the patch from the upstream fixes this properly and elegantly, while my patch makes the code return the opposite result, which makes it impossible for torchaudio to adopt the upstream libsox codebase.

So I do not recommend applying my patch. I plan to get rid of this in torchaudio by migrating the in-memory decoding/encoding capability to FFmpeg.

I do not know the policy of condo-forge about BC-breaking, but once you apply my patch, you will have technical debt hard to fix.

Reply:
Thanks for the response!

> I think the patch from the upstream fixes this properly and elegantly, while my patch makes the code return the opposite result, which makes it impossible for torchaudio to adopt the upstream libsox codebase.

I don't understand the need for divergence (if their solution is elegant, why couldn't it work for torchaudio? If it doesn't work for torchaudio, how can it be fixed properly?)

> I do not know the policy of conda-forge about BC-breaking, but once you apply my patch, you will have technical debt hard to fix.

Not sure what you mean with BC, maybe binary compatibility? We shouldn't break except where upstream does so as well (e.g. major releases). Thanks for the heads up. I might have to bite the bullet and take the vendored sox for torchaudio then until you get rid of it (IIUC)...

Reply:
> I don't understand the need for divergence (if their solution is elegant, why couldn't it work for torchaudio? If it doesn't work for torchaudio, how can it be fixed properly?)

I debugged the issue myself and created the patch tailored for torchaudio's need, and after a while I realized that the upstream had fix which would not work for torchaudio.

> Not sure what you mean with BC

Applying the patch from upstream will fix the issue of `is_seekable()` returning random value, and it will return `false` for in-memory audio data. If you apply my patch, then `is_seekable()` will return true for in-memory audio data. If you apply my patch, and when torchaudio gets rid of my patch, libsox from condo-forge will be the only one with the custom behavior, which I think is something you want to get rid of at some point. Then getting rid of my patch incurs backward-incompatible change, which will disturb user experience.

Reply:
well, so far we don't have torchaudio at all, so there's no user experience to break yet.

And IIUC, you're going to keep the torchaudio-specific behaviour also when you switch to ffmpeg as a backend?

However, the sox builds in conda-forge (with an "a" 😉) are not just used by torchaudio, so we have to keep that in mind.

Reply:
This leaves three options:
* do nothing --> users stay exposed to the hard to debug behaviour
* use the torchaudio-fix --> should be an improvement, but they might get used to the "wrong" behaviour (assuming sox ever does another release, last one was 7.5 years ago...)
* use the upstream fix --> incompatible with torchaudio, but compatible with possible future sox releases (also means we have to vendor lots of stuff into torchaudio, which is also unappealing).

They all have their pros & cons TBH. Not sure if anyone else from @conda-forge/sox has an opinion on this?